### ISSUE TITLE
No keyboard detected on Dell Optiplex 580, no SPD data

### ISSUE THREAD

**Issue by 01e3** [memtest86plus/memtest86plus#71]
I have a Dell Optiplex 580 DT where keyboard does not work on Memtest86+. It is connected to a dell KVM via a dongle, but just in case - I also plugged a "real" keyboard. Memtest86+ is booted in BIOS via LILO - this machines does not support UEFI. I also tested the official USB/ISO "GRUB" image - no changes.

Also, there is no SPI data but I think this may be WAI given the ATI/AMD SMBus controller that may not be yet supported. Just a note that SPD works on Linux:

```
[    9.003538] piix4_smbus 0000:00:14.0: SMBus Host Controller at 0xb00, revision 0
[    9.008553] piix4_smbus 0000:00:14.0: Auxiliary SMBus Host Controller at 0xb20

# i2cdetect -l|grep SMBus
i2c-6   smbus           SMBus PIIX4 adapter at 0b20             SMBus adapter
i2c-5   smbus           SMBus PIIX4 adapter at 0b00             SMBus adapter

# i2cdetect -y 5
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: 30 31 32 33 -- -- -- -- 38 -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- 69 -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

# decode-dimms|grep ^Number
Number of SDRAM DIMMs detected and decoded: 4
```

Attaching the screenshots and the output from lspci/lsusb/lsusb.py:

[dello580-lsusb.py.txt](https://github.com/memtest86plus/memtest86plus/files/8701488/dello580-lsusb.py.txt)
[dello580-lsusb.txt](https://github.com/memtest86plus/memtest86plus/files/8701489/dello580-lsusb.txt)
[dello580-lspci.txt](https://github.com/memtest86plus/memtest86plus/files/8701490/dello580-lspci.txt)

![no-keyboard-Dell-Optiplex-580-1](https://user-images.githubusercontent.com/105136494/168631526-4eb6a39c-de7e-44fb-b67d-f0c9087ff2c2.png)
![no-keyboard-Dell-Optiplex-580-2](https://user-images.githubusercontent.com/105136494/168631538-2d3fddc2-17db-4c54-bbd3-c62cdfc2dce6.png)

**Comment by x86fr** [comment]
About SPD, your chipset is on top of the list for addition.

**Comment by debrouxl** [comment]
* tracking issue for PIIX4: #69 ;
* the USB controllers are the same as those in #6 :)

**Comment by martinwhitaker** [comment]
To confirm this is the same as issue #6, try booting with the "noehci" option.

**Comment by 01e3** [comment]
noehci does work!

**Comment by martinwhitaker** [comment]
Does this patch make any difference?
````
--- a/system/ehci.c
+++ b/system/ehci.c
@@ -649,6 +649,7 @@ bool ehci_init(int bus, int dev, int func, uintptr_t base_addr, usb_hcd_t *hcd)
     if (num_keyboards == 0) {
         // Halt the host controller.
         (void)halt_host_controller(op_regs);
+        (void)reset_host_controller(op_regs);
 
         // Deallocate the workspace for this controller.
         pm_map[0].end += num_pages(sizeof(workspace_t));
````

**Comment by 01e3** [comment]
Oh, it does!

Will double check tomorrow that I did not mess up anything with noehci or something and provide the screenshot.

Observation: I wonder if it would make sense to report the options passed to memtest?

**Comment by debrouxl** [comment]
Same for me, the keyboard now works on the RS904A, which only supports legacy BIOS boot :)
The USB keyboard operation is undisturbed on both the PowerEdge R815 (same AMD chipset) in UEFI mode and the GA-M61P-S3 (NVIDIA MCP61) in legacy BIOS mode. I don't think I'll have time to make regression tests on more computers until the week-end, though.

**Comment by 01e3** [comment]
Happy Dell Optiplex 580 with keyboard working (usbdebug):

![no-keyboard-Dell-Optiplex-580-3](https://user-images.githubusercontent.com/105136494/169206436-74f4e69b-30f0-4c3a-abc2-6310ba920e21.png)

**Comment by x86fr** [comment]
> Does this patch make any difference?
> 
> ```
> --- a/system/ehci.c
> +++ b/system/ehci.c
> @@ -649,6 +649,7 @@ bool ehci_init(int bus, int dev, int func, uintptr_t base_addr, usb_hcd_t *hcd)
>      if (num_keyboards == 0) {
>          // Halt the host controller.
>          (void)halt_host_controller(op_regs);
> +        (void)reset_host_controller(op_regs);
>  
>          // Deallocate the workspace for this controller.
>          pm_map[0].end += num_pages(sizeof(workspace_t));
> ```

About ~50% of the motherboards I just tested (AMD 2010-2015 era) have this issue with keyboard undetected in some or all USB ports. This fix solve it in all cases. Can you commit it to main?

**Comment by martinwhitaker** [comment]
Fixed by commit 3a10701.
